zilliqa/src/consensus.rs

```diff
@@ -1887,27 +1840,47 @@ impl Consensus {
         Ok(())
     }
 
-    pub fn leader(&self, committee: &Committee, view: u64) -> Validator {
+    pub fn leader_at_block(&self, block: &Block, view: u64) -> Option<Validator> {
+        let Ok(state_at) = self.try_get_state_at(block.number()) else {
+            return None;
+        };
+        Some(self.leader(&state_at, view))
+    }
+
+    pub fn leader(&self, state: &State, view: u64) -> Validator {
+        let committee = state.get_stakers().unwrap();
+
         let mut rng = ChaCha8Rng::seed_from_u64(view);
-        let dist = WeightedIndex::new(committee.iter().map(|v| {
-            let stake = self
-                .state
-                .get_stake(v.public_key)
+        let dist = WeightedIndex::new(committee.iter().map(|pub_key| {
+            let stake = state
+                .get_stake(*pub_key)
                 .unwrap()
                 .context("Committee member has no stake")
                 .unwrap();
             stake.get()
         }))
         .unwrap();
         let index = dist.sample(&mut rng);
-        committee.iter().nth(index).unwrap()
+        let public_key = *committee.get(index).unwrap();
```

Comment:
Maybe another performance point for the future - We could avoid a 2nd EVM call here if `getStakers` returned the `(pubKey, rewardAddress, peerId, stake, etc)` instead of just the pub key?